Importing Libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Dataset from Kaggle => https://www.kaggle.com/code/janiobachmann/credit-fraud-dealing-with-imbalanced-datasets/input

Reading the Dataset

In [2]:
df = pd.read_csv("C:/Users/Hp/Downloads/codsoft internship/creditcard.csv")
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

Checking the Null values

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [4]:
df.isnull().sum().sum()

0

In [5]:
df["Class"].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

Making 2 different datasets on basis of "0"(legit) and "1"(fraud).

In [6]:
df_legit = df[df["Class"]==0]
df_fraud = df[df["Class"]==1]

In [7]:
df_legit["Amount"].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [8]:
df_fraud["Amount"].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [9]:
df.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Making a new dataset which containsequal number of class as the "fraud" count is '492' and "legit" count is '284315'.

In [10]:
legit_sample = df_legit.sample(n= 492)

In [11]:
new_sample = pd.concat([legit_sample, df_fraud], axis=0)

This is our new datset which will going to use further for appyling models.

In [12]:
new_sample

Time         V1         V2         V3        V4        V5  \
207323  136620.0  -0.900551  -1.482331   0.092161 -1.952743 -1.280788   
140902   84010.0  -7.028598 -13.665820  -3.921644  4.663142 -4.641473   
49776    44208.0 -25.374386  -4.438911 -13.388378  5.930693 -7.287802   
260013  159367.0  -0.603920   0.330019   1.750916 -2.604553 -0.473440   
16622    27981.0  -2.303672  -1.660566   1.338367  0.757566  2.489735   
...          ...        ...        ...        ...       ...       ...   
279863  169142.0  -1.927883   1.125653  -4.518331  1.749293 -1.566487   
280143  169347.0   1.378559   1.289381  -5.004247  1.411850  0.442581   
280149  169351.0  -0.676143   1.126366  -2.213700  0.468308 -1.120541   
281144  169966.0  -3.113832   0.585864  -5.399730  1.817092 -0.840618   
281674  170348.0   1.991976   0.158476  -2.583441  0.408670  1.151147   

              V6         V7         V8        V9  ...        V21       V22  \
207323 -0.552726   1.697750  -0.518750 -2.883558  ...   0.037970 -0.226244   
140902  1.979854   5.156590  -0.973013  0.489855  ...   2.349991 -2.627204   
49776   1.544321 -10.381358 -13.155712 -1.375833  ... -11.502155  3.410590   
260013 -0.585334   0.566259  -0.381646  1.797717  ...   0.250690  1.142751   
16622  -0.860350  -0.221569   0.049707 -0.653040  ...   0.089882  0.490220   
...          ...        ...        ...       ...  ...        ...       ...   
279863 -2.010494  -0.882850   0.697211 -2.064945  ...   0.778584 -0.319189   
280143 -1.326536  -1.413170   0.248525 -1.127396  ...   0.370612  0.028234   
280149 -0.003346  -2.234739   1.210158 -0.652250  ...   0.751826  0.834108   
281144 -2.943548  -2.208002   1.058733 -1.632333  ...   0.583276 -0.269209   
281674 -0.096695   0.223050  -0.068384  0.577829  ...  -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28   Amount  \
207323  0.841420 -0.000051  0.249591 -0.164085 -0.044932  0.155372   435.00   
140902 -3.872036 -0.344018 -0.980779 -0.853732 -0.731338  0.770651  4276.82   
49776   2.613389  1.013962  0.277105  0.002981  4.379080 -8.338123     1.00   
260013 -0.660412 -0.030751  0.793503 -0.438036 -0.281374 -0.165678    25.95   
16622   0.836313 -0.267619  0.062135  0.468201 -0.210371  0.037873     0.89   
...          ...       ...       ...       ...       ...       ...      ...   
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968   390.00   
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637     0.76   
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361    77.89   
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700   245.00   
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309    42.53   

        Class  
207323      0  
140902      0  
49776       0  
260013      0  
16622       0  
...       ...  
279863      1  
280143      1  
280149      1  
281144      1  
281674      1  

[984 rows x 31 columns]

In [13]:
new_sample["Class"].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [14]:
new_sample.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92972.465447 -0.103476 -0.039143  0.095600 -0.051905  0.065272   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.034225 -0.002495 -0.000273 -0.055845  ... -0.024653 -0.053122   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.066345  0.016984  0.010579 -0.042130  0.009120  0.019407 -0.021668   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       83.291707  
1      122.211321  

[2 rows x 30 columns]

Spliting the new datset into train and test. 

In [15]:
x = new_sample.drop(columns=["Class"], axis= 1)
y = new_sample.iloc[:,-1]

In [16]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.25, random_state=42)

Now, will apply the model i.e; Logistic Regression.

In [17]:
lr = LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

Will see the accuracy of both: 'train' and 'test' data; to see whether the model is "proper" or "overfitting" or "underfitting".

In [18]:
train_pred = lr.predict(x_train)
train_acc = accuracy_score(train_pred, y_train)
train_acc

0.9214092140921409

In [19]:
test_pred = lr.predict(x_test)
test_acc = accuracy_score(test_pred, y_test)
test_acc

0.9024390243902439

The model is "proper".